File: codes/094.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}
def inorder_traversal(root)
    result = []
    inorder(root, result)
    return result
end

def inorder(node, result)
    return if node.nil?
    inorder(node.left, result) if node.left
    result << node.val
    inorder(node.right, result) if node.right
end

File: codes/094.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
    stack := []*TreeNode{}
    res:=[]int{}

    for root!=nil || len(stack) != 0{
        if root.Left!=nil{
            stack = append(stack,root)
            root = root.Left
        }else{
            res = append(res,root.Val)
            root = root.Right
            for root==nil && len(stack) !=0{
                root = stack[len(stack)-1]
                stack = stack[:len(stack)-1]
                res = append(res,root.Val)
                root = root.Right
            }
        }
    }
    return res
}
